epilogue-processor/src/main/java/edu/wpi/first/epilogue/processor/LoggerGenerator.java

```diff
@@ -0,0 +1,251 @@
+            .map(e -> (VariableElement) e)
+            .filter(notSkipped)
+            .filter(optedIn)
+            .filter(e -> !e.getModifiers().contains(Modifier.STATIC))
+            .filter(this::isLoggable)
+            .toList();
+
+    var methodsToLog =
+        clazz.getEnclosedElements().stream()
+            .filter(e -> e instanceof ExecutableElement)
+            .map(e -> (ExecutableElement) e)
+            .filter(notSkipped)
+            .filter(optedIn)
+            .filter(e -> !e.getModifiers().contains(Modifier.STATIC))
+            .filter(e -> e.getModifiers().contains(Modifier.PUBLIC))
+            .filter(e -> e.getParameters().isEmpty())
+            .filter(e -> e.getReceiverType() != null)
+            .filter(this::isLoggable)
+            .toList();
+
+    writeLoggerFile(clazz.getQualifiedName().toString(), epilogue, fieldsToLog, methodsToLog);
+  }
+
+  private void writeLoggerFile(
+      String className,
+      Epilogue classConfig,
+      List<VariableElement> loggableFields,
+      List<ExecutableElement> loggableMethods)
+      throws IOException {
+    String packageName = null;
+    int lastDot = className.lastIndexOf('.');
+    if (lastDot > 0) {
+      packageName = className.substring(0, lastDot);
+    }
+
+    String simpleClassName = StringUtils.simpleName(className);
+    String loggerClassName = className + "Logger";
+    String loggerSimpleClassName = loggerClassName.substring(lastDot + 1);
+
+    // Use the name on the class config to set the generated logger names
+    // This helps to avoid naming conflicts
+    if (!classConfig.name().isBlank()) {
+      loggerSimpleClassName =
+          StringUtils.capitalize(classConfig.name().replaceAll(" ", "")) + "Logger";
+      if (lastDot > 0) {
+        loggerClassName = packageName + "." + loggerSimpleClassName;
+      } else {
+        loggerClassName = loggerSimpleClassName;
+      }
+    }
+
+    var loggerFile = m_processingEnv.getFiler().createSourceFile(loggerClassName);
+
+    var privateFields =
+        loggableFields.stream().filter(e -> e.getModifiers().contains(Modifier.PRIVATE)).toList();
+    boolean requiresVarHandles = !privateFields.isEmpty();
+
+    try (var out = new PrintWriter(loggerFile.openWriter())) {
+      if (packageName != null) {
+        // package com.example;
+        out.println("package " + packageName + ";");
+        out.println();
+      }
+
+      out.println("import edu.wpi.first.epilogue.Epilogue;");
+      out.println("import edu.wpi.first.epilogue.Epiloguer;");
+      out.println("import edu.wpi.first.epilogue.logging.ClassSpecificLogger;");
+      out.println("import edu.wpi.first.epilogue.logging.DataLogger;");
+      if (requiresVarHandles) {
+        out.println("import java.lang.invoke.MethodHandles;");
+        out.println("import java.lang.invoke.VarHandle;");
+      }
+      out.println();
+
+      // public class FooLogger implements ClassSpecificLogger<Foo> {
```

Comment:
whoops

Reply:
Hah. I originally had this printing out elements individually, so comments were helpful for showing examples of what the code was generating. I later refactored it to use string concats and this got left over